### PR TITLE
KAFKA-9375 Add thread names to kafka connect

### PR DESCRIPTION
**Pull request by cryptoe**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by cryptoe** [comment]
@ryannedolan / @mimaison Can you please review this too ?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java [hunk 1]
```diff
@@ -16,6 +16,7 @@
  */
 package org.apache.kafka.connect.runtime.distributed;
 
+import com.google.common.util.concurrent.Runnables;
```
Comment (gcsaba2) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java]: Are you sure this is going to work? I don't see Guava in dependencies.gradle and can't find any com.google packages anywhere in the code.

Comment (cryptoe) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java]: My bad . Seems my local intellij needed some cleanup. Thanks for the catch.

**Comment by mimaison** [comment]
Thanks for the PR, I'll try to review it later this week.

File: clients/src/test/java/org/apache/kafka/common/utils/ThreadUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        public void run() {
+        }
+    };
+    private static final String THREAD_NAME = "ThreadName";
+    private static final String THREAD_NAME_WITH_NUMBER = THREAD_NAME + "%d";
+
+
+    @Test
+    public void testThreadNameWithoutNumberNoDemon() {
+        Assert.assertEquals(ThreadUtils.createThreadFactory(THREAD_NAME, false).
```
Comment (mimaison) [inline clients/src/test/java/org/apache/kafka/common/utils/ThreadUtilsTest.java]: In all `assert` method, the first argument is the "expected" value while the second one is the "actual" value. Can you swap them in this file to match that definition?

File: clients/src/test/java/org/apache/kafka/common/utils/ThreadUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        Thread daemonThread = ThreadUtils.createThreadFactory(THREAD_NAME, true).newThread(EMPTY_RUNNABLE);
+        try {
+            Assert.assertEquals(daemonThread.getName(), THREAD_NAME);
+            Assert.assertTrue(daemonThread.isDaemon());
+        } finally {
+            try {
+                daemonThread.join();
+            } catch (InterruptedException e) {
+                // can be ignored
+                e.printStackTrace();
```
Comment (mimaison) [inline clients/src/test/java/org/apache/kafka/common/utils/ThreadUtilsTest.java]: Maybe we don't need to print the stack trace if we can ignore it

File: clients/src/test/java/org/apache/kafka/common/utils/ThreadUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        ThreadFactory localThreadFactory = ThreadUtils.createThreadFactory(THREAD_NAME_WITH_NUMBER, false);
+        Assert.assertEquals(localThreadFactory.newThread(EMPTY_RUNNABLE).getName(), THREAD_NAME + "1");
+        Assert.assertEquals(localThreadFactory.newThread(EMPTY_RUNNABLE).getName(), THREAD_NAME + "2");
+    }
+
+    @Test
+    public void testThreadNameWithNumberDemon() {
+        ThreadFactory localThreadFactory = ThreadUtils.createThreadFactory(THREAD_NAME_WITH_NUMBER, true);
+        Thread daemonThread1 = localThreadFactory.newThread(EMPTY_RUNNABLE);
+        Thread daemonThread2 = localThreadFactory.newThread(EMPTY_RUNNABLE);
```
Comment (mimaison) [inline clients/src/test/java/org/apache/kafka/common/utils/ThreadUtilsTest.java:66]: Do we need `daemonThread2` in this test?

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/FileOffsetBackingStoreTest.java [hunk 5]
```diff
@@ -20,6 +20,7 @@
 import org.apache.kafka.connect.util.Callback;
 import org.easymock.EasyMock;
 import org.junit.After;
+import org.junit.Assert;
```
Comment (mimaison) [inline connect/runtime/src/test/java/org/apache/kafka/connect/storage/FileOffsetBackingStoreTest.java]: Can we import static like `assertEquals`?

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/FileOffsetBackingStoreTest.java [hunk 6]
```diff
@@ -41,6 +43,11 @@
     File tempFile;
 
     private static Map<ByteBuffer, ByteBuffer> firstSet = new HashMap<>();
+    private static final Runnable EMPTY_RUNNABLE = new Runnable() {
```
Comment (mimaison) [inline connect/runtime/src/test/java/org/apache/kafka/connect/storage/FileOffsetBackingStoreTest.java]: I believe you can use the Java 8 syntax: `Runnable runnable = () -> {};`. This also appears in another file

**Review by mimaison — CHANGES_REQUESTED** [review]
Thanks for the PR! It's a nice improvement. I've left a few comments

File: clients/src/test/java/org/apache/kafka/common/utils/ThreadUtilsTest.java (hunk 3 above)
Comment (cryptoe) [reply clients/src/test/java/org/apache/kafka/common/utils/ThreadUtilsTest.java]: Sure. Will do that

File: clients/src/test/java/org/apache/kafka/common/utils/ThreadUtilsTest.java (hunk 2 above)
Comment (cryptoe) [reply clients/src/test/java/org/apache/kafka/common/utils/ThreadUtilsTest.java]: acked

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/FileOffsetBackingStoreTest.java (hunk 6 above)
Comment (cryptoe) [reply connect/runtime/src/test/java/org/apache/kafka/connect/storage/FileOffsetBackingStoreTest.java]: Will do that

File: clients/src/test/java/org/apache/kafka/common/utils/ThreadUtilsTest.java (hunk 4 above)
Comment (cryptoe) [reply clients/src/test/java/org/apache/kafka/common/utils/ThreadUtilsTest.java:66]: I wanted to check if 2 daemonThreads are appropriately numbered.

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/FileOffsetBackingStoreTest.java (hunk 5 above)
Comment (cryptoe) [reply connect/runtime/src/test/java/org/apache/kafka/connect/storage/FileOffsetBackingStoreTest.java]: Done

**Comment by cryptoe** [comment]
@mimaison Thanks for looking into this. I have updated the PR based on the review comments.

**Review by ryannedolan — APPROVED** [review]
lgtm, thx

**Review by mimaison — APPROVED** [review]
LGTM

**Comment by mimaison** [comment]
retest this please

**Comment by mimaison** [comment]
Failures look unrelated, merging

**Pull request merged**